src/native/libs/System.Security.Cryptography.Native/pal_ssl.c

```diff
@@ -586,6 +599,61 @@ void CryptoNative_SslCtxSetCaching(SSL_CTX* ctx, int mode)
     {
         SSL_CTX_set_options(ctx, SSL_OP_NO_TICKET);
     }
+
+    if (newSessionCb != NULL || removeSessionCb != NULL)
+    if (newSessionCb != NULL)
+    {
+        SSL_CTX_sess_set_new_cb(ctx, newSessionCb);
+    }
+
+    if (removeSessionCb != NULL)
+    {
+        SSL_CTX_sess_set_remove_cb(ctx, removeSessionCb);
+    }
+
+    return retValue;
+}
+
+const char* CryptoNative_SslGetServerName(SSL * ssl)
+{
+    return SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name);
+}
+
+int32_t CryptoNative_SslSetSession(SSL* ssl, SSL_SESSION *session)
+{
+    return SSL_set_session(ssl, session);
+}
+
+void CryptoNative_SslSessionFree(SSL_SESSION *session)
+{
+    SSL_SESSION_free(session);
+}
+
+const char* CryptoNative_SslSessionGetHostname(SSL_SESSION *session)
+{
+#ifdef SSL_SESSION_get0_hostname
```

Comment:
I don't believe the `ifdef` will evaluate to true on a non-portable build.  (Since it's a function identifier then)

This should probably instead be based on the NEED_OPENSSL_x_y values.  

Try building with `./build-native.sh -portableBuild=false` and see if it works as expected.

Reply:
I did build on Ubuntu 16.04 with OpenSSL 1.0.1e with portableBuild `true` and `false`. And it did not fail. Same on Ubuntu 20.04 with OpenSSL 1.1.1. 
Would NEED_OPENSSL_1_1 cover also 3.0? I can still change that if you prefer it.

Reply:
> I did build on Ubuntu 16.04 with OpenSSL 1.0.1e with portableBuild true and false. And it did not fail. 

That's... really surprising.  Looking at the headers over time, it looks like it was added in 1.1.0, and has always been a function (vs a macro acting like a function), so what I'd expect is

1.0.1e in non-portable:

The compile doesn't see any definition of SSL_SESSION_get0_hostname at all, turns this into `#if false` and removes the code.

1.1.1 in non-portable:

SSL_SESSION_get0_hostname is a function, so the preprocessor doesn't see it, so this still turns into `#if false` and goes away.

In both cases the compile would succeed, but I'd expect 1.1.1-direct to fail to run (remember that `./build-native.sh doesn't update the testhost, so you need to manually copy over it or make it a symlink -- or just load up the compile output in a debugger and see if there's a function body or not)

> Would NEED_OPENSSL_1_1 cover also 3.0?

No, you'd need `#if defined NEED_OPENSSL_1_1 || defined NEED_OPENSSL_3_0`

Reply:
you are right. updated. It is interesting that we don't really have _any_ test coverage for non-portable builds, right? (aside from compilation)